estimator.go

```diff
@@ -0,0 +1,343 @@
+	callData, err := estimator.Encode("estimate", call.To, call.Data)
+	if err != nil {
+		return nil, err
+	}
+
+	type Call struct {
+		To   common.Address `json:"to"`
+		Data string         `json:"data"`
+	}
+
+	estimateCall := &Call{
+		To:   from,
+		Data: "0x" + common.Bytes2Hex(callData),
+	}
+
+	var res string
+	err = provider.RPC.Call(&res, "eth_call", estimateCall, blockTag, finalOverwrites)
+	if err != nil {
+		return nil, err
+	}
+
+	resBytes := common.Hex2Bytes(strings.Replace(res, "0x", "", 1))
+
+	var success bool
+	var result []byte
+	var gas *big.Int
+
+	if err := ethcoder.AbiDecoder([]string{"bool", "bytes", "uint256"}, resBytes, []interface{}{&success, &result, &gas}); err != nil {
+		return nil, err
+	}
+
+	gas.Add(gas, big.NewInt(int64(e.CalldataCost(call.Data))))
+
+	if !success {
+		reason := string(result[68 : len(result)-1])
+		return gas, fmt.Errorf("error calling restimate: " + reason)
+	}
+
+	return gas, nil
+}
+
+func (e *Estimator) AreEOAs(ctx context.Context, provider *ethrpc.Provider, walletConfig WalletConfig) ([]bool, error) {
+	// Get non-eoa signers
+	// required for computing worse case scenario
+	// TODO: Do these calls in parallel
+	isEoa := make([]bool, walletConfig.Signers.Len())
+	for i, s := range walletConfig.Signers {
+		code, err := provider.CodeAt(ctx, s.Address, nil)
+		if err != nil {
+			return nil, err
+		}
+
+		isEoa[i] = len(code) == 0
+	}
+
+	return isEoa, nil
+}
+
+func (e *Estimator) PickSigners(ctx context.Context, walletConfig WalletConfig, isEoa []bool) ([]bool, error) {
+	type SortedSigner struct {
+		s WalletConfigSigner
+		i int
+	}
+
+	// Create a copy of the signers array
+	// this will be sorted and used to pick the worst case scenario for the signers
+	sortedSigners := make([]SortedSigner, walletConfig.Signers.Len())
+	for i, _ := range walletConfig.Signers {
+		sortedSigners[i] = SortedSigner{
+			s: walletConfig.Signers[i],
+			i: i,
+		}
+	}
+
+	sort.SliceStable(sortedSigners, func(a, b int) bool {
+		if !isEoa[sortedSigners[a].i] && isEoa[sortedSigners[b].i] {
+			return true
+		}
+
+		return sortedSigners[a].s.Weight < sortedSigners[b].s.Weight
+	})
+
+	weightSum := 0
+
+	// Define what signers are goint go be signing
+	// it should construct a worse case scenario for the signature
+	willSign := make([]bool, walletConfig.Signers.Len())
+	threshold := int(walletConfig.Threshold)
+
+	// Pick signers until we reach the threshold we stop
+	// We use the sorted signers to get the ones with the non EOA and with lowest weight first
+	for _, s := range sortedSigners {
+		weightSum += int(s.s.Weight)
+		willSign[s.i] = true
+
+		if weightSum >= threshold {
+			break
+		}
+	}
+
+	return willSign, nil
+}
+
+func stubAddress() common.Address {
+	raw := make([]byte, 20)
+	rand.Read(raw)
+	return common.BytesToAddress(raw)
+}
+
+func (e *Estimator) BuildStubSignature(walletConfig WalletConfig, willSign []bool, isEoa []bool) []byte {
```

Comment:
Just a thought, but what if instead of building the stub signature in the estimator, we have the wallets themselves be able to construct their own stub signatures since they're likely more aware of what signers they will be using?

Reply:
Relayer would still need it's own independent logic because the wallet could lie about the using the worst case scenario.

Reply:
Yep exactly, but this may not be a bad idea for the nested signers problem.

The wallet could provide the Config for all nested signers, then the relayer could validate if the provided config matches the configuration of those signers.

But it would add more complexity to the whole thing, I think it's better to keep it simple for now.

Reply:
But if the wallet lies, it would be doing it to its own detriment, no? The wallet would lie, saying it expects to sign with a really short signature, then sign with a really long signature, which would fail the relayer's validation check.